### PR TITLE
Add support for git remote resolution using provider APIs with an API token

### PR DESCRIPTION
**Pull request by abayer**
# Changes

Closes https://github.com/tektoncd/resolution/issues/42

This adds support for API-based token authentication and fetching of remote resources without requiring a full clone of the repository to the `git` resolver, using https://github.com/jenkins-x/go-scm for interacting with provider APIs. It is fully backwards compatible with the existing `git` resolver behavior and syntax, with the new behavior being used when the `repo` parameter is passed to the `git` resolver instead of `url`.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Adds new behavior to the git remote resolver to fetch Tasks and Pipelines from GitHub, Gitlab, BitBucket via their APIs using API tokens.
```

### CODE REVIEW

**Comment by lbernick** [comment]
dumb question, SCM = source control manager right? How does a user know when to use the git resolver vs this one?

**Comment by vdemeester** [comment]
> > dumb question, SCM = source control manager right? How does a user know when to use the git resolver vs this one?
> 
> Correct. Maybe it should be renamed "git-api"? In retrospect, I wish the git resolver had been named something different, like "anonymous-git", because the most important difference is that the scm resolver authenticates (using an API key), so it can access private repos. It also fetches files via the provider's API, so it doesn't need to do a full clone, etc.

- We can still rename 👼🏼 (before doing the release)
- Is that resolver support more than just `git` ? (and is the library we use for this support more than git ?) Depending on the error, we may want to rename this one with `git` in the name or not. `git-provider` would make more sense than `git-api` for me (as.. it is configurable and "tied" to a git provider, or a scm provider)

**Comment by abayer** [comment]
`git-provider` does feel more right to me. Can anyone think of a shorter name for the existing non-authenticated, full-clone provider than `anonymous-git`?

**Comment by vdemeester** [comment]
I meant `git-provider` for this resolver, not the anonymous one

**Comment by abayer** [comment]
So did I. =)

So here's what I'm saying:
* The existing `git` resolver gets renamed to `anonymous-git`, or hopefully a better name than that.
* This new resolver gets renamed to `git-provider`.

**Comment by lbernick** [comment]
Could we just have one git resolver that does not require credentials, but uses them if they are present?

**Comment by abayer** [comment]
> Could we just have one git resolver that does not require credentials, but uses them if they are present?

The problem/challenge there is that their mechanisms and params differ - that said, I guess we could have it use the API if SCM provider/creds/org/repo are provided, and anonymous/full clone if they aren't and a URL is provided.

**Comment by lbernick** [comment]
> > Could we just have one git resolver that does not require credentials, but uses them if they are present?
> 
> The problem/challenge there is that their mechanisms and params differ - that said, I guess we could have it use the API if SCM provider/creds/org/repo are provided, and anonymous/full clone if they aren't and a URL is provided.

I think that would be ideal if it's workable

**Comment by abayer** [comment]
Ok, first attempt at merging the functionality is up - lots of cleanup to do tomorrow, assuming it works.

**Comment by abayer** [comment]
/hold cancel

Ok, I have cleaned up my work from yesterday, removed the separate `scm` resolver, and updated docs etc accordingly. This should be good to go.

**Comment by vicpadilla** [comment]
Has this approach been considered for fetching a single file?: https://stackoverflow.com/questions/28375418/git-how-to-pull-a-single-file-from-a-server-repository-in-git
It would allow us to continue using the go-git library which is provider-agnostic.

The go-git library also seems to support authentication: https://stackoverflow.com/a/48895644/2570297

**Comment by abayer** [comment]
@vicpadilla - That has not - but it also looks like it still needs a full clone of the repo, and seems to be about updating a single file in a clone, not just fetching a single file without cloning, sadly. Could you open an issue for the go-git authentication? Thanks!

File: config/resolvers/config-feature-flags.yaml
Comment (lbernick) [inline config/resolvers/config-feature-flags.yaml:44]: I think this can be removed

Comment (abayer) [reply config/resolvers/config-feature-flags.yaml:44]: Whoops, I missed a reference! Fixing.

File: pkg/resolution/resolver/git/resolver.go
Comment (lbernick) [inline pkg/resolution/resolver/git/resolver.go:34]: curious why you're using jenkins-x/go-scm instead of https://github.com/drone/go-scm (the one linked to in the docs)?

File: test/git-resolver/gitea.yaml
Comment (lbernick) [inline test/git-resolver/gitea.yaml:1]: what's this file for?

File: pkg/apis/config/resolver/feature_flags.go
Comment (abayer) [reply pkg/apis/config/resolver/feature_flags.go:37]: oops!

File: pkg/resolution/resolver/git/resolver.go
Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:101]: That was a mistake, putting it back.

File: pkg/resolution/resolver/git/resolver.go
Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:34]: I link to the drone one? Lemme fix that, should be jenkins-x. That's the one I _know_ works, since I spent months working on it. =)

File: test/git-resolver/gitea.yaml
Comment (abayer) [reply test/git-resolver/gitea.yaml:1]: Spinning up gitea in the cluster for the git API resolution e2e tests. I'll add a comment.

File: pkg/resolution/resolver/git/resolver.go
Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:101]: They'll be at the bottom of `populateDefaultParams` now.

File: pkg/resolution/resolver/framework/testing/featureflag.go
Comment (lbernick) [inline pkg/resolution/resolver/framework/testing/featureflag.go:50]: this can be removed too

**Review by lbernick — APPROVED** [review]
thanks Andrew-- approving in the sense of "this looks reasonable and has tests" but not in the sense of "I'm able to approve every line for correctness". maybe one change that would help would be splitting the gitea server creation into a separate PR

File: pkg/resolution/resolver/framework/testing/featureflag.go
Comment (abayer) [reply pkg/resolution/resolver/framework/testing/featureflag.go:50]: hahaha, how do I keep missing these? =)

**Comment by abayer** [comment]
fwiw, I'm planning to move the gitea setup stuff to plumbing after this has merged, as well as making it a bit more generic. There are other places that could take advantage of having tooling for setting up gitea, like https://github.com/tektoncd/plumbing/pull/1185, for example. So I think it's fine to rubber-stamp the gitea stuff for now and give it a more thorough review when I move it to plumbing.

File: pkg/resolution/resolver/git/resolver.go
Comment (chuangw6) [inline pkg/resolution/resolver/git/resolver.go:307]: Is it possible to explicitly set the annotation keys to be namespaced directly in `annotations.go` rather than doing it at time of writing the annotations i.e. `AnnotationKeyPath=internal.BuiltInAnnotation("path")`? Doing it here implicitly might mislead that the annotation key is just original `AnnotationKeyPath` rather than the name-spaced.

Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:307]: Yeah, that could be done, I just got tired of copy-pasting `resolution.tekton.dev/` everywhere. =)

Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:307]: ...and I'm gonna do it. First a PR to update all the existing resolvers other than `git` (since that'll be dealt with in this PR), then updating this PR.

Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:307]: Oh, wait, I added it in this PR in the first place! 🤦

Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go]: aaaand done.

File: config/resolvers/git-resolver-config.yaml
Comment (vdemeester) [inline config/resolvers/git-resolver-config.yaml:28]: I think I missed this earlier 🤔 Why do we default to the catalog repo ?

File: docs/git-resolver.md
Comment (vdemeester) [inline docs/git-resolver.md:13]: It's `url` or `repo`+`org` right ? Not sure how to make it clear in the parameter list but we should try 👼🏼

**Review by vdemeester — COMMENTED** [review]
On question : there would have to be one resolver per provider, am I right ? If on my cluster I want to be able to fetch task/pipeline from both `github` and my internal `gitlab` (or even gitlab.com), I would need to run 2 resolvers with different names ? or is it even possible ? (I would love it if it was 👼🏼 )

**Comment by abayer** [comment]
> On question : there would have to be one resolver per provider, am I right ? If on my cluster I want to be able to fetch task/pipeline from both `github` and my internal `gitlab` (or even gitlab.com), I would need to run 2 resolvers with different names ? or is it even possible ? (I would love it if it was 👼🏼 )

As of now, yes, we can only support one provider. I need to think more about how we can support multiple providers/credentials/etc, but I think that can come later.

File: config/resolvers/git-resolver-config.yaml
Comment (abayer) [reply config/resolvers/git-resolver-config.yaml:28]: I have no idea, but it was already that way when I copied it over from resolution.git. =)

File: docs/git-resolver.md
Comment (abayer) [reply docs/git-resolver.md:13]: Yeah, the wording could be improved, but I haven't come up with better yet.

File: pkg/resolution/resolver/git/resolver.go
Comment (abayer) [reply pkg/resolution/resolver/git/resolver.go:129]: 👍

**Comment by vdemeester** [comment]
/lgtm
We can figure out the multiple providers later on as this is still alpha 👼🏼

**Pull request merged**
